allennlp/training/model.py

```diff
@@ -0,0 +1,64 @@
+from typing import Dict
+import torch
+
+
+class Model(torch.nn.Module):
+    """
+    This abstract class represents a model to be trained. Rather than relying completely
+    on the Pytorch Module, we provide a richer API that maintains more metadata
+    about the tensors used. This makes reading model implementations easier to
+    connect to data generation, because model inputs are named.
+
+    Models built using this API are still compatible with other pytorch models and can
+    be used naturally as modules within other models - outputs are dictionaries, which
+    can be unpacked and passed into other layers. One caveat to this is that if you
+    wish to use an AllenNLP model inside a Container (such as nn.Sequential), you must
+    interleave the models with a wrapper module which unpacks the dictionary into
+    a list of tensors.
+
+    Finally, in order for your model to be trained using the :class:`~allennlp.training.Trainer`
+    api, the output dictionary of your Model must include a "loss" key, which will be
+    optimised during the training process.
+    """
+    #  TODO(Mark): Implement this.
+    def forward(self, *inputs) -> Dict[str, torch.Tensor]:  # pylint: disable=arguments-differ
```

Comment:
It also saves unpacking the dictionary inside your forward call and makes it explicit that the `labels` are optional, as they can have `None` default arguments.